Exclude sites from sensor totals and Energy dashboard

autoSteve opened the pull request:
This may get legs.

Initial dev. deems it quite simple to implement, and low risk for the mainstream, despite being a feature that hardly anyone will use. Which is good. Less issue reports. 🫢

BJReplay commented:
Lol - saw the MI reference in the email notification :)

autoSteve commented:
FYI. We're still what the checklist considers Plat.

MyPy strict type checking result:

```
(ha-venv) vscode ➜ /workspaces/…/homeassistant/components/solcast_solar $ ./mypy_test.sh 
Success: no issues found in 6 source files
Success: no issues found in 1 source file
Success: no issues found in 4 source files
(ha-venv) vscode ➜ /workspaces/…/homeassistant/components/solcast_solar $
```

PyTest coverage:

```
---------- coverage: platform linux, python 3.12.8-final-0 -----------
Name                                                                                     Stmts   Miss  Cover   Missing
----------------------------------------------------------------------------------------------------------------------
/workspaces/homeAssistant-core/homeassistant/components/solcast_solar/__init__.py          390      0   100%
/workspaces/homeAssistant-core/homeassistant/components/solcast_solar/config_flow.py       229      0   100%
/workspaces/homeAssistant-core/homeassistant/components/solcast_solar/const.py              43      0   100%
/workspaces/homeAssistant-core/homeassistant/components/solcast_solar/coordinator.py       287      0   100%
/workspaces/homeAssistant-core/homeassistant/components/solcast_solar/diagnostics.py        17      0   100%
/workspaces/homeAssistant-core/homeassistant/components/solcast_solar/energy.py             15      0   100%
/workspaces/homeAssistant-core/homeassistant/components/solcast_solar/repairs.py            41      0   100%
/workspaces/homeAssistant-core/homeassistant/components/solcast_solar/select.py             44      0   100%
/workspaces/homeAssistant-core/homeassistant/components/solcast_solar/sensor.py            150      0   100%
/workspaces/homeAssistant-core/homeassistant/components/solcast_solar/solcastapi.py       1461      0   100%
/workspaces/homeAssistant-core/homeassistant/components/solcast_solar/system_health.py      11      0   100%
/workspaces/homeAssistant-core/homeassistant/components/solcast_solar/util.py              125      0   100%
----------------------------------------------------------------------------------------------------------------------
TOTAL                                                                                     2813      0   100%

Results (29.70s):
      69 passed
```

I think this is good to go, but I'll leave it in draft until further confirmation of usability. Mixed messages overnight, and all that.

autoSteve commented:
Positive feedback finally for site exclusion.

BJReplay commented:
SLGTM 🤷

gcoan commented on README.md:
Comments from me:

"It allows forecast visualisation in the Energy dashboard" reword to "It allows visualisation of the solar forecast in the Energy dashboard"

---

"Solcast will produce a forecast of solar PV generation from today through seven days into the future" to "Solcast will produce a forecast of solar PV generation from today through to seven days into the future"

---

"The Energy dashboard in Home Assistant is populated with a specific data structure that is provided by the integration. (This does not come from sensor data.) Manipulation of forecasted values to account for shading is possible by setting dampening factors for hourly or half-hourly periods, and a "hard limit" may be set for over-sized solar arrays where expected generation cannot exceed an inverter maximum rating. These two mechanisms are the only ways to manipulate the data.

Historical forecast data is retained by the integration for up to two years, and this is stored in an integration data file. Forecast history is not stored as Home Assistant statistics."

to

"The Energy dashboard in Home Assistant is populated with a specific historical forecast data file that is provided by the integration, and is retained for up to two years. 

Manipulation of forecasted values to account for shading is possible by setting dampening factors for hourly or half-hourly periods, and a "hard limit" may be set for over-sized solar arrays where expected generation cannot exceed an inverter maximum rating. These two mechanisms are the only ways to manipulate the Solcast forecast data.

Forecast history is not stored as Home Assistant statistics."

---

"You need to restart Home Assistant before you can then install the custom component that you've just downloaded."

to

"You need to restart Home Assistant before you can then configure the Solcast PV Forecast custom component that you've just downloaded."

---

"Enter your Solcast API Key, API limit, desired auto-update choice and click Submit. If you have more than one Solcast account because you have more than two rooftop setups, enter both account API keys separated by a comma xxxxxxxx-xxxxx-xxxx,yyyyyyyy-yyyyy-yyyy (Note: this goes against Solcast T&C's by having more than one account). If the API limit is the same for multiple accounts then enter a single value for that, or both values separated by a comma."

to

"Enter your Solcast API Key, API limit, desired auto-update choice and click Submit. If you have more than one Solcast account because you have more than two rooftop setups, enter both Solcast account API keys separated by a comma xxxxxxxx-xxxxx-xxxx,yyyyyyyy-yyyyy-yyyy (Note: this goes against Solcast T&C's by having more than one account). Your API limit will be 10 for new Solcast users or 50 for historical users. If the API limit is the same for multiple accounts then enter a single value for that, or both values separated by a comma."

---

"Edit the Solar Panels Solar production item you have previously created (or will create now). Do not add a separate Solar production item as things will just get weird."

to

"The solar forecast has to be associated with a solar generation item in your Energy dashboard.

Edit a Solar Panels Solar production item you have previously created (or will create now). Do not add a separate Solar production item as things will just get weird.

There can only be a single configuration of the total Solcast PV Forecast in the Energy dashboard, it is not possible to split the forecast for different solar arrays/Solcast sites."

---

"All sensor names are preceded by Solcast PV Forecast."

to

"All sensor names are preceded by the integration name, Solcast PV Forecast."

---

"As stated above, sensor attributes are created to enable sensor state variations to be used in templates. Examples are the estimate confidence, estimate10/estimate50/estimate90. The sensor state is generally left at the default of estimate50, but displaying the tenth percentile of a sensor on a dashboard may be desired and this is enabled by the use of attribute values."

to

"As stated above, sensor attributes are created to enable sensor state variations to be used in templates. Examples are the estimate confidence, estimate10/estimate/estimate90. The sensor state is generally left at the default of estimate, but displaying the tenth percentile of a sensor on a dashboard may be desired and this is enabled by the use of attribute values."

And all other references to 'estimate50' (including in the JSON) should refer to 'estimate'

---

Question:

"Rooftop site name attributes include:

name: The site name configured at solcast.com.
resource_id: The site resource ID."

should it be the site resource id with hyphens replaced by underscores so can use it directly in an attribute retrieval?

although this may impact dampening configuration, so probably can't win here

---

"You can change the dampening factor value for any hour. Values from 0.0 - 1.0 are valid. Setting 0.95 will dampen each Solcast forecast data value by 5%. This is reflected in the sensor values and attributes and also in the Home Assistant Energy dashboard."

to

"You can change the dampening factor value for any hour. Values from 0.0 - 1.0 are valid. Setting 0.95 will dampen (reduce) each Solcast forecast data value by 5%. This is reflected in the sensor values and attributes and also in the Home Assistant Energy dashboard."

---

"action: solcast_solar.query_forecast_data
data:
  start_date_time: 2024-10-08T12:00:00+11:00
  end_date_time: 2024-10-08T19:00:00+11:00
  undampened: true
  #site: 1111-aaaa-bbbb-2222"

is the site supposed to be commented out in the example service call?

---

"If granular dampening is set to specify both individual site factors and an 'all' factors, then attempting retrieval of an individual site factors will result in the 'all' factors being returned, with the 'all' site being noted in the response. This is because an 'all' set of factors overrides the individual site settings in this circumstance."

to

"If granular dampening is set to specify both individual site dampening factors and 'all sites' dampening factors, then attempting retrieval of an individual site dampening factors will result in the 'all sites' dampening factors being returned, with the 'all' site being noted in the response. This is because an 'all' set of dampening factors overrides the individual site settings in this circumstance."

---

autoSteve replied:
Great feedback, Geoff.

I was going to write something like what did your last slave die of, can't you commit all this yourself? 😂 But then realised that this branch is on @BJReplay's repo and I am a collab yet you are not.

Ugh.

Yeah, the comment in `YAML` is intended. That param is optional, and the comment underscores it. "#" is generally accepted as a comment thing, but I do get your point.

Hypens vs. underscores? I'm not going there... It irks me that I did not know that about attribute naming in the first place. #smartereveryday. It'd be possible to do a string replace in a Jinja2 template if needed. Should it be needed? Nope. Sorry everybody "advanced" for adding to tech-legacy. Rip the bandaid off and change it? I'm chickening out on that one for now, but good thought.

gcoan replied:
Thanks Steve, sorry you had to re-apply all my changes, but I wasn't a formal reviewer so don't think I could have made them directly.  Did try to make it easier by giving the full original and replacements so you can search and replace; and I did think about highlighting the changes in bold but then decided against it as it would have made the merge more ucky.

Yes the hyphens vs underscores for site id attribute in the rooftop entity is inconsistent; I suspect almost no-one uses it anyway, and it is fixable in Jinja should you ever need to use it.
And changing site id's for dampening factors would be a breaking change for those that use dampening factors,  so lets not go there.

Anyway, it is what it is, if we had our time again things would be different, but they're not so we live with it.

Thanks

BJReplay reviewed (approved):
I preferred the bot template 🤣

BJReplay reviewed (commented):
Reviewing from phone, so hard to be thorough: did I miss tests?

autoSteve commented:
> Reviewing from phone, so hard to be thorough: did I miss tests?

I can add assertions for the cleaned up entities, if that's what you mean. That is currently not confirmed by PyTest, but does get 100% code path coverage.

BJReplay commented:
> Now tested

I am sure it is fine, but I didn't spot anything and just wanted to mention it.

Just got home, very tired, will look again tomorrow, but am sure it is all good.

autoSteve commented:
And there's more to come.

Evolution. I re-built my dev. container, and the HA PyTest set up in there now validates that string things are in strings.json.

Handy.

But ugh. 😩 Enjoyment of listening to The Bloodhound Gang on vinyl halted...

Config/options flow tests are now updated to cope. Translations to go.

The skinny is that the config/options error strings were always presented by us in English and not translated. That translation appears to be mandatory now from a PyTest POV.

"Mandatory" also needs to be worked around now. Like if you're passing through exception text, like a 429 response from solcast for a specific named API key, as I do. So I had to code exceptions, which is not a very elegant/performant thing from our perspective for one of the tests.

It is what it is. I've added a full second of runtime to the full suite of tests...

autoSteve commented:
I'm slipping... It's on the to-do list.

```
---------- coverage: platform linux, python 3.13.3-final-0 -----------
Name                                                                                     Stmts   Miss  Cover   Missing
----------------------------------------------------------------------------------------------------------------------
/workspaces/homeAssistant-core/homeassistant/components/solcast_solar/__init__.py          403      0   100%
/workspaces/homeAssistant-core/homeassistant/components/solcast_solar/config_flow.py       231      0   100%
/workspaces/homeAssistant-core/homeassistant/components/solcast_solar/const.py              43      0   100%
/workspaces/homeAssistant-core/homeassistant/components/solcast_solar/coordinator.py       290      0   100%
/workspaces/homeAssistant-core/homeassistant/components/solcast_solar/diagnostics.py        17      0   100%
/workspaces/homeAssistant-core/homeassistant/components/solcast_solar/energy.py             15      0   100%
/workspaces/homeAssistant-core/homeassistant/components/solcast_solar/repairs.py            41      0   100%
/workspaces/homeAssistant-core/homeassistant/components/solcast_solar/select.py             44      0   100%
/workspaces/homeAssistant-core/homeassistant/components/solcast_solar/sensor.py            158      0   100%
/workspaces/homeAssistant-core/homeassistant/components/solcast_solar/solcastapi.py       1461      0   100%
/workspaces/homeAssistant-core/homeassistant/components/solcast_solar/system_health.py      11      0   100%
/workspaces/homeAssistant-core/homeassistant/components/solcast_solar/util.py              125      4    97%   94-95, 97, 99
----------------------------------------------------------------------------------------------------------------------
TOTAL                                                                                     2839      4    99%
```

```
(ha-venv) vscode ➜ /workspaces/…/homeassistant/components/solcast_solar $ ./mypy_test.sh 
energy.py:33: error: Item "None" of "ConfigEntry[Any] | None" has no attribute "runtime_data"  [union-attr]
config_flow.py:108: error: Name "dt_util.dt.tzinfo" is not defined  [name-defined]
config_flow.py:162: error: Argument 1 of "async_get_options_flow" is incompatible with supertype "ConfigFlow"; supertype defines the argument type as "ConfigEntry[Any]"  [override]
config_flow.py:162: note: This violates the Liskov substitution principle
config_flow.py:162: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides
config_flow.py:177: error: Incompatible types in assignment (expression has type "ConfigEntry[Any] | None", variable has type "SolcastConfigEntry")  [assignment]
config_flow.py:218: error: Incompatible types in assignment (expression has type "ConfigEntry[Any] | None", variable has type "SolcastConfigEntry")  [assignment]
Found 5 errors in 2 files (checked 6 source files)
select_mod.py:19: error: Module "homeassistant.helpers.entity" does not explicitly export attribute "DeviceInfo"  [attr-defined]
Found 1 error in 1 file (checked 1 source file)
solcastapi.py:462: error: Redundant cast to "dict[Any, Any]"  [redundant-cast]
solcastapi.py:910: error: Argument 1 to "async_update_entry" of "ConfigEntries" has incompatible type "SolcastConfigEntry | None"; expected "ConfigEntry[Any]"  [arg-type]
solcastapi.py:3060: error: Dict entry 1 has incompatible type "str": "SolcastConfigEntry"; expected "str": "str | int | float | None"  [dict-item]
sensor.py:31: error: Module "homeassistant.helpers.entity" does not explicitly export attribute "EntityCategory"  [attr-defined]
Found 4 errors in 2 files (checked 4 source files)
```

Stupid evolution.

BJReplay commented:
> The skinny is that the config/options error strings were always presented by us in English and not translated. That translation appears to be mandatory now from a PyTest POV.

I've finally got back to this. Long story (via a signal group chat - I'll include a senior editor of an appropriate journal).

autoSteve commented:
Six nasty strict type checking issues to go that actually don't mean shit. Python is by design not a strict type checking language, so these are "nasty" from the point of view that I need to fool MyPy into thinking that something kosher is happening.

The code is doing exactly what it needs to do. MyPy just hates me.

The code runs and tests perfectly.

```
energy.py:33: error: Item "None" of "ConfigEntry[Any] | None" has no attribute "runtime_data"  [union-attr]
config_flow.py:178: error: Argument 1 to "SolcastSolarOptionFlowHandler" has incompatible type "ConfigEntry[Any]"; expected "SolcastConfigEntry"  [arg-type]
config_flow.py:182: error: Incompatible types in assignment (expression has type "ConfigEntry[Any] | None", variable has type "SolcastConfigEntry")  [assignment]
config_flow.py:223: error: Incompatible types in assignment (expression has type "ConfigEntry[Any] | None", variable has type "SolcastConfigEntry")  [assignment]
solcastapi.py:910: error: Argument 1 to "async_update_entry" of "ConfigEntries" has incompatible type "SolcastConfigEntry | None"; expected "ConfigEntry[Any]"  [arg-type]
sensor.py:31: error: Module "homeassistant.helpers.entity" does not explicitly export attribute "EntityCategory"  [attr-defined]
```

BJReplay commented:
> I need to get smarter, clearly, as this code did parse correctly before I upgraded my dev. container so it is frustrating.

I had something similar for a while when I first started using inherited config entries in my EPA Air Quality integration.  ruff complaining about a type error - but it still worked.  

And then it was gone.

If I was to guess - and all I can do is guess :) - I would guess that it is either a change in the dev container definition or a change in the HA dev environment. 

When I re-open my EPA AQ dev container now... (waiting, waiting, waiting)

I see this which makes me think that the HA team have changed the definition

![image](https://github.com/user-attachments/assets/94f731f7-adfb-4f4a-83b3-fcfc450fd449)

Since I am a risk-taking guy (I'm not, but these are bits, and I can get new ones for free), I click on rebuild.

Anyway, it's not complaining now with this in `__init__.py`

``` python
from homeassistant.config_entries import ConfigEntry
from homeassistant.core import HomeAssistant
from homeassistant.exceptions import ConfigEntryNotReady

_LOGGER = logging.getLogger(__name__)

PLATFORMS = [Platform.SENSOR]

type EPAConfigEntry = ConfigEntry[EPAData]
```

the `type` definition used to give me a big red squiggle.

I don't know if this helps, or this is the annoying _I've seen something similar, it must be the same_.

autoSteve commented:
Thanks for the thoughts.

I did away with the SolcastConfigEntry type definition altogether, and needed to put a few type ignore directives in places where the type checking was never going to pass. Four of these: `(expression has type "ConfigEntry[Any] | None", variable has type "ConfigEntry[Any]")`.

You'd think I could just type the variables as `entry: ConfigEntry | None`, but that caused other havoc that was never going to pass.

```
(ha-venv) vscode ➜ /workspaces/…/homeassistant/components/solcast_solar $ ./mypy_test.sh
Success: no issues found in 6 source files
Success: no issues found in 1 source file
Success: no issues found in 4 source files

---------- coverage: platform linux, python 3.13.3-final-0 -----------
Name                                                                                     Stmts   Miss  Cover   Missing
----------------------------------------------------------------------------------------------------------------------
/workspaces/homeAssistant-core/homeassistant/components/solcast_solar/__init__.py          407      0   100%
/workspaces/homeAssistant-core/homeassistant/components/solcast_solar/config_flow.py       232      0   100%
/workspaces/homeAssistant-core/homeassistant/components/solcast_solar/const.py              43      0   100%
/workspaces/homeAssistant-core/homeassistant/components/solcast_solar/coordinator.py       290      0   100%
/workspaces/homeAssistant-core/homeassistant/components/solcast_solar/diagnostics.py        17      0   100%
/workspaces/homeAssistant-core/homeassistant/components/solcast_solar/energy.py             16      0   100%
/workspaces/homeAssistant-core/homeassistant/components/solcast_solar/repairs.py            42      0   100%
/workspaces/homeAssistant-core/homeassistant/components/solcast_solar/select.py             43      0   100%
/workspaces/homeAssistant-core/homeassistant/components/solcast_solar/sensor.py            157      0   100%
/workspaces/homeAssistant-core/homeassistant/components/solcast_solar/solcastapi.py       1462      0   100%
/workspaces/homeAssistant-core/homeassistant/components/solcast_solar/system_health.py      11      0   100%
/workspaces/homeAssistant-core/homeassistant/components/solcast_solar/util.py              128      0   100%
----------------------------------------------------------------------------------------------------------------------
TOTAL                                                                                     2848      0   100%


Results (26.42s):
      74 passed
```

autoSteve commented:
Just one of these left that is suppressed: `expression has type "ConfigEntry[Any] | None", variable has type "ConfigEntry[Any]"`.

I've traced the root to a Home Assistant function that does not have a compatible positional parameter type hint, so this last one likely won't be fixable.

autoSteve commented:
All code is now type-checked, even the Flask-based simulator.

This is almost 100% strict checking for the integration and sim. (there is one suppressed line for the integration as noted earlier), and it's a little lax in places for PyTest modules. (Lax in tests for code clarity and brevity mostly, but despite the laxity the tests will still blow up if anything is amiss. The test suppressions prevent VScode lighting up with reported module problem noise.)

I've installed the MyPy extension in my dev. container, so now I get near real-time alerts about type checking. Interestingly the VScode extension provides superior type checking over what the command-line mypy delivers. Maybe the extension is further evolved. 🤷‍♂️

autoSteve commented:
Integration and sim. are now both 100% strict, with no suppression.

autoSteve commented:
I'm going to merge this, BJ.

Reckon we could send it on Friday the 2nd May. I've not got much planned for that weekend except for smashing shit up followed by lunch with a mate. https://thesmashroom.com.au/.

BJReplay commented:
> I'm going to merge this, BJ.

Send it.

> smashing shit up

Enjoy!

Many years ago, I had the pleasure of disposing of some dead monitors before e-wasting properly was the done thing: the skip was in an alleyway, and we figured we could probably get the monitors into the skip from the roof that was the equivalent of level 3.  A couple of ~collaborators~ colleagues blocked off either end of the alleyway (safety, first, of course), and we lugged the monitors up a level to the roof instead of down a level and, at first carefully tipped them over the edge.

The most satisfying impacts were monitor on monitor as they both explode and implode at the same time.
